common/txmgr/tracker.go

```diff
@@ -0,0 +1,295 @@
+package txmgr
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	feetypes "github.com/smartcontractkit/chainlink/v2/common/fee/types"
+	txmgrtypes "github.com/smartcontractkit/chainlink/v2/common/txmgr/types"
+	"github.com/smartcontractkit/chainlink/v2/common/types"
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+
+	"github.com/smartcontractkit/chainlink/v2/core/utils"
+)
+
+const (
+	// defaultTTL is the default time to live for abandoned transactions
+	// After this TTL, the TXM stops tracking abandoned Txs.
+	defaultTTL = 6 * time.Hour
+)
+
+// AbandonedTx is a transaction who's 'FromAddress' was removed from the KeyStore(by the Node Operator).
+// Thus, any new attempts for this Tx can't be signed/created. This means no fee bumping can be done.
+// However, the Tx may still have live attempts in the chain's mempool, and could get confirmed on the
+// chain as-is. Thus, the TXM should not directly discard this Tx.
+type AbandonedTx[ADDR types.Hashable] struct {
+	id          int64
+	fromAddress ADDR
+}
+
+// Tracker tracks all transactions which have abandoned fromAddresses.
+// The fromAddresses can be deleted by Node Operators from the KeyStore. In such cases,
+// existing in-flight transactions for these fromAddresses are considered abandoned too.
+// Since such Txs can still have attempts on chain's mempool, these could still be confirmed.
+// This tracker just tracks such Txs for some time, in case they get confirmed as-is.
+type Tracker[
+	CHAIN_ID types.ID,
+	ADDR types.Hashable,
+	TX_HASH types.Hashable,
+	BLOCK_HASH types.Hashable,
+	R txmgrtypes.ChainReceipt[TX_HASH, BLOCK_HASH],
+	SEQ types.Sequence,
+	FEE feetypes.Fee,
+] struct {
+	txStore      txmgrtypes.TxStore[ADDR, CHAIN_ID, TX_HASH, BLOCK_HASH, R, SEQ, FEE]
+	lggr         logger.Logger
+	enabledAddrs map[ADDR]bool
+	txCache      map[int64]AbandonedTx[ADDR]
+	ttl          time.Duration
+	lock         sync.Mutex
+	started      bool
+	mb           *utils.Mailbox[int64]
+	chDone       chan struct{}
+	wg           sync.WaitGroup
+}
+
+func NewTracker[
+	CHAIN_ID types.ID,
+	ADDR types.Hashable,
+	TX_HASH types.Hashable,
+	BLOCK_HASH types.Hashable,
+	R txmgrtypes.ChainReceipt[TX_HASH, BLOCK_HASH],
+	SEQ types.Sequence,
+	FEE feetypes.Fee,
+](
+	txStore txmgrtypes.TxStore[ADDR, CHAIN_ID, TX_HASH, BLOCK_HASH, R, SEQ, FEE],
+	lggr logger.Logger,
+) *Tracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE] {
+	return &Tracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]{
+		txStore:      txStore,
+		lggr:         lggr.Named("Tracker"),
+		enabledAddrs: map[ADDR]bool{},
+		txCache:      map[int64]AbandonedTx[ADDR]{},
+		ttl:          defaultTTL,
+		mb:           utils.NewSingleMailbox[int64](),
+		chDone:       make(chan struct{}),
```

Comment:
Using Confirmer as an example, maybe we can replace this with a `ctx` and `ctxCancel`. That way we'll have a context for the txstore calls we do in other methods and can call cancel on it during close. Otherwise passing a `context.Background()` without a cancel wouldn't properly reclaim resources when things are shutting down.